Geometry configuration check (issue #24328) [1/10]

PetrilloAtWork opened the pull request:
This is part of the pull request set including also LArSoft/larcoreobj#9.
Breaking changes need to be resolved by the experiments: the resolution for LArSoft is in LArSoft/larcorealg#12, LArSoft/lardataalg#18, LArSoft/lardata#12, LArSoft/larevt#10, LArSoft/larsim#56, LArSoft/larreco#24, LArSoft/larexamples#5 _(edited 20201207)_ and LArSoft/lareventdisplay#10 _(edited 20201214)_
The branches offered for evaluation to the experiments, all named `feature/gp_issue24328`, are in the following repositories:
* ArgoNeuT: `argoneutcode`
* DUNE: `dunetpc`
* ICARUS: `sbncode`, `icarusalg`, `icaruscode`
* SBND: `sbncode`, `sbndcode`
* MicroBooNE: `uboonecode`, `ubana`, `ubcrt`, `ubcv`, `ubevt`, `ublite`, `ubraw`, `ubreco`, `ubsim`
_(edited 20201214)_

This is the resolution of [Redmine issue #24328](https://cdcvs.fnal.gov/redmine/issues/24328).
The main discussion on the issue resolution will happen in this pull request.

A refresher of how the solution is deployed is in the new documentation of `geo::Geometry` class. That text is the reference design of the solution.

The feature has been "manually" tested by providing crafted inputs to configurations all including `Geometry` service and, unless otherwise specified, `GeometryConfigurationWriter` service:
* no input file (source: `EmptyEvent`): geometry information added to output file
* empty input file, with a job not including the new `GeometryConfigurationWriter` service: configuration error
* empty input file: geometry information added to output file
* input file with `sumdata::RunData` data product compatible with current `Geometry`: geometry information added to output file
* input file with `sumdata::RunData` data product not compatible with current `Geometry`: exception showing the different configurations
* input file with `sumdata::GeometryConfigurationInfo` data product compatible with current `Geometry`: **data product declared but not put**; but the job still succeeds as required
* input file with `sumdata::GeometryConfigurationInfo` data product not compatible with current `Geometry`: exception showing the different configurations

Further action is needed to address the data product declared but not put (because one already existed). I have discussed with @knoepfel who confirmed that _art_ is currently not enforcing the insertion of a declared data product in the context that this feature uses, and that we can go along with that and fix it in an unlikely future where this check becomes enforced.

## Breaking change and resolution

The new feature _requires_ the experiments to run `GeometryConfigurationWriter` service when an input file which has never seen `Geometry` service is processed with a job that sets `Geometry` service up. The easiest (and safe) way to do that is to configure `GeometryConfigurationWriter` together with `Geometry` and `ExptGeoHelperInterface`, and run it every time.
The recommended way to configure the geometry was something like:
```
experiment_geometry_services: {
  Geometry:               @local::experiment_geometry
  ExptGeoHelperInterface: @local::experiment_geometry_helper
}
```
with the user configurations including something like:
```
services: {
  @table::experiment_geometry_services
  # ...
}
```
In this scenario, the configuration fix is to change the `experiment_geometry_services` configuration table as follows:
```
experiment_geometry_services: {
  Geometry:               @local::experiment_geometry
  ExptGeoHelperInterface: @local::experiment_geometry_helper
  GeometryConfigurationWriter: {}
}
```
In the cases where a geometry configuration bundle is not used, the line `services.GeometryConfigurationWriter: {}` or equivalent must be added to the configuration.

If there are reasons to wish to skip the check, `Geometry` service can be instructed to do so, with an option like:
```
services.Geometry.SkipConfigurationCheck: true
```
**Note**: the service `GeometryConfigurationWriter` must be configured even if the check is skipped with `SkipConfigurationCheck` set to `true`.

PetrilloAtWork committed:
Replaced geometry reload on new run with configuration compatibility check.

This is part of the resolution of issue #24328.
This is also a breaking change: `GeomertyConfigurationWriter` service is now required when using `Geometry` service.

FNALbuild commented:
A new Pull Request was created by  @PetrilloAtWork (Gianluca Petrillo) for develop.

It involves the following packages:

larcore


@LArSoft/level-1-managers, @LArSoft/level-2-managers can you please review it and eventually sign? Thanks.

cms-bot commands are listed <a href="https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Pull_request_comments_that_trigger_CI_actions">here</a>

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//722/console

PetrilloAtWork committed:
Alternative implementation of GeometryConfigurationWriter.

The configuration is always written.

FNALbuild commented:
Pull request #6 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//725/console

PetrilloAtWork commented:
The current design has an issue in case geometry configuration information is already present in the input `art::Run` (i.e. if the service has already been run when the input file was created).
The  `GeometryConfigurationWriter` service can't know beforehand whether there is already the necessary information in the runs or not, so it always has to promise it `produces()` it.
I can see three options for `GeometryConfigurationWriter` service:

1. the service may or may not fulfil the promise depending on the content of each run; this is the first implementation I have pushed; despite what the documentation of `ProducingService` says, `produces()` promise seems to be not checked, so that implementation passed the tests; it appears from the test that when asking for a `GeometryConfigurationInfo` data product, the ones that are not present (like if they were dropped) are ignored, and _art_ keeps trying until it finds one which is there.
2. the service will always produce information; if some already exists, it will duplicate it again; this is the "alternative" implementation I have pushed. The issue is the growing of duplicate information (at least, it will be consistent!).
3. the service always produces a collection of information objects, either empty or with a single element (or an `std::optional` if _art_ and ROOT support it); the drawback is a cumbersome implementation with a data collection which is not a collection, and a complex reading step where all the collections of geometry information objects must be read by type, selected by label (but not by process name) to include only data products produced by past instances of the service, and then checking one (or all) of the instances that are not empty and ignoring the others.
4. a variation of the first design would be to tell _art_ that we changed our mind and to treat the data product that we promised as dropped, since we are not going to put it in; I don't know if this is supported by _art_.

Advise is welcome.

PetrilloAtWork committed:
Added unit tests for geometry version check.

FNALbuild commented:
Pull request #6 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//764/console

PetrilloAtWork committed:
Added geometry check unit test.

This one is with check disabled and no GeometryConfigurationWriter.
Should still fail.

FNALbuild commented:
Pull request #6 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//766/console

knoepfel commented on larcore/Geometry/Geometry_service.cc:
```diff
@@ -162,5 +165,98 @@ namespace geo {
+  {
+    
+    try {
+      return run.getByLabel<sumdata::GeometryConfigurationInfo>
+        (art::InputTag{"GeometryConfigurationWriter"});
```
Should this be a `getValidHandle` call?  I expect that `getByLabel` would never fail unless there is corrupt data.

PetrilloAtWork replied:
If the question is whether `getByLabel()` (overload with a single input tag as argument) throws when there is no product, from the unit tests it appears to do it, big time.
I did not use `getValidHandle()` because here I don't care about the handle but about the data product (hint is that if I used `getValidHandle()`, I would need further processing to `return` what I need to).
I might be missing your point though.

knoepfel replied:
Ah, I missed that the one-argument `getByLabel` overload was being used.  Sorry about that.

knoepfel commented on larcore/Geometry/GeometryConfigurationWriter_service.cc:
```diff
@@ -0,0 +1,254 @@
+    {
+      return std::make_unique<sumdata::GeometryConfigurationInfo>
+        (std::forward<Args>(args)...);
+    }
```
Suggest making this a normal (perhaps `static`) member function--all uses of it below pass one argument and it's always of the same type.  Avoiding unnecessary templates prevents unnecessary compilation time.

PetrilloAtWork replied:
It is not exactly true that the type in the three calls is the same: in one case is a l-value `const` reference, and in the other two it is a r-value one.
We are hitting the border of my understanding of C++ here. Is it correct that the same effect can be obtained by passing the argument by value?
```
static InfoPtr_t makeInfoPtr(sumdata::GeometryConfigurationInfo info)
    {
      return std::make_unique<sumdata::GeometryConfigurationInfo>
        (std::move(info));
    }
```
In my understanding, passing a r-value reference would move its content into `info`, and passing a l-value reference would copy it.

knoepfel replied:
You're not getting a lot of savings by using `std::move` on a `GeometryConfigurationInfo` object--it contains a number and two strings.  So in this case, I wouldn't worry about eking out those 10s of bytes saved by using the move in one case and the copy in the other.  The code will be easier to read, and almost as performance to just do:
```.cpp
static InfoPtr_t makInfoPtr(sumdata::GeometryConfiguationInfo const& info)
{
  return std::make_unique<sumdata::GeometryConfigurationInfo>(info);
}
```
In fact, it may be _just as_ performant as you might be copying the stack's info object to the heap anyway (depending on the size of the internal strings).

knoepfel commented on larcore/Geometry/GeometryConfigurationWriter_service.cc:
```diff
@@ -0,0 +1,254 @@
+  /// Writes the information from the service configuration into the `run`.
+  virtual void postReadRun(art::Run& run) override;
+  
+  // -- END --- Interface ------------------------------------------------------
```
Why is this labeled "Interface"?  This service has no publicly accessible interface.

PetrilloAtWork replied:
I did not mean interface between class implementation and class caller, but rather between class and base class.
That block was supposed to enclose all base class methods that would be overridden.
Is the term "interface" inappropriate? Feel free to propose a better one.

knoepfel replied:
Using the term `override` is already self-documenting--you can only override a base class membee function--thus no additional commentary is necessary.  You could say, "BEGIN --- Overrides ...", but then you're just repeating what's already apparent.

PetrilloAtWork replied:
Those markers are for fast lookup rather than for documentation.
I have replaced them.

knoepfel reviewed (commented):
Looks excellent, Gianluca.  I post a few comments below for your consideration.

PetrilloAtWork committed:
Fixed some typos and an unnecessary header inclusion.

PetrilloAtWork committed:
Geometry information from legacy RunData has its own version now

FNALbuild commented:
Pull request #6 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//781/console

PetrilloAtWork committed:
Adopted some suggestions from review.

PetrilloAtWork committed:
Merge remote-tracking branch 'upstream/develop' into feature/gp_issue24328

FNALbuild commented:
Pull request #6 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//783/console

PetrilloAtWork commented:
I have adopted some of the suggestions.
I have also updated the code to `develop` branches. Which did not take much effort since there were no changes.

knoepfel commented:
@PetrilloAtWork, I have marked this PR as ready for review.  Assuming L2 approval, I will also approve this PR.  Would you would be willing to present a summary of these changes at the next LArSoft Coordination Meeting on Dec. 1?

absolution1 commented:
approve

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged.)

FNALbuild commented:
-LArSoft tests failed on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/10189&builds= 
 -build

FNALbuild commented:
-LArSoft tests failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/10188&builds= 
 -build

absolution1 commented:
Hiya.  I'm guessing the reason for the failures is that CI tests for this CI need to pull in other PRs as well?

lgarren commented:
@PetrilloAtWork  We triggered the CI with all 3 PRs.  A number of larsoft unit tests are failing with this error:
```
87: %MSG-s ArtException:  PostEndJob 30-Nov-2020 11:43:21 CST ModuleEndJob
88: ---- EventProcessorFailure BEGIN
89:   EventProcessor: an exception occurred during current event processing
90:   ---- ProductNotFound BEGIN
91:     Can't read geometry configuration information.
92:     Is `GeometryConfigurationWriter` service configured?
93:     
94:     ---- ProductNotFound BEGIN
95:       getBySelector: Found zero products matching all criteria
96:       Looking for type: sumdata::GeometryConfigurationInfo
97:     ---- ProductNotFound END
98:   ---- ProductNotFound END
99: ---- EventProcessorFailure END
100: %MSG
101: Art has completed and will exit with status 1.
```
See https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/10219&builds=

PetrilloAtWork commented:
I am not surprised, given that this is a breaking change.
The procedure for the update is described in the [first post](https://github.com/LArSoft/larcore/pull/6#issue-499646386).
I do not have enough understanding of the configuration strategy for each experiment to be able to apply the change in the needed configurations. Help is appreciated.

knoepfel commented:
@PetrilloAtWork, the failures are not in experiment code.  Look at (e.g.) `larevt`, and go from there.

PetrilloAtWork commented:
I have misread the information Lynn had provided.
I'll check.

PetrilloAtWork committed:
Fix to unit test.

PetrilloAtWork committed:
Configuration updates related to issue #24328

FNALbuild commented:
Pull request #6 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//800/console

PetrilloAtWork commented:
I have changed a number of configurations that contain `Geometry` service within LArSoft.
A few of them indeed involved unit tests. Most of them involve configurations that are broken for other aspects.
And I did not even touch the ones requiring experiment configuration (mostly DUNE, a few MicroBooNE), which should probably not be in LArSoft at all (or maybe they are meant as examples?).
The result is additional six pull requests, which are listed in the first post of this thread (which I have just updated).

I proceeded by looking for all FHiCL configuration files which contain `Geometry:` and manually editing them.
I verified that unit tests on `larsoft_suite` and `larsoftobj_suite` pass with `e19:prof` setup.

lgarren commented:
e19 CI results
https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/10285&builds=
problem in argoneutcode unit tests
https://dbweb0.fnal.gov/LarCI/app/ns:ArgoNeuT/view_builds/index?offset=0&builds=argoneut_ci/5349&builds=
problem in SBND unit tests
https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/5840&builds=
problem in ICARUS unit tests
https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/1435&builds=
dunetpc CI failures
https://dbweb0.fnal.gov/LarCI/app/ns:DUNE/view_builds/index?offset=0&builds=dune_ci/10611&builds=
problem in uboone unit tests
https://dbweb0.fnal.gov/LarCI/app/ns:uBooNE/view_builds/index?offset=0&builds=uboone_ci/5770&builds=

lgarren commented:
This is a great step forward, but we have to understand and provide fixes for experiment code as well.

PetrilloAtWork commented:
I reiterate the request for support, and advise.
There are ~15 configurations in `argoneutcode` which do not follow the recommendation of having a geometry service configuration bundle, and may need a fix¹; likewise, about 250 from `dunetpc`; 5 in `lariatsoft`; 4 in `sbncode`/`sbndcode`; and ~60 in `uboone_suite`. It is not clear to me which of them are used and which are obsolete. The fix can't be safely automated, so the fix needs to happen file by file.
How should I proceed? I am comfortable to fix the unit test configurations. The C.I. too (although I am not completely sure they'll succeed). I might figure out the central configurations (things like `services_argoneut.fcl`).
That is _a lot_, and it leaves _even more_ out.
Please advise.

---- 

¹ I should point out that adhering to the recommendation to use a configuration table ("bundle") including `Geometry` and `ExptGeometryHelperInterface` does not exempt from the fix, it just makes that fix much straightforward.

lgarren commented:
@yangtj207 @tomjunk  Are you able to help @PetrilloAtWork ?

tomjunk commented:
So is the recommended maintenance to add

GeometryConfigurationWriter: {}

to configuration files everywhere Geometry and ExptGeoHelperInterface is defined?
I only see 58 such instances in the top-level dunetpc:

grep -r -i exptgeohelperinterface * | grep fcl | wc -l
58
 
perhaps I am being naive and am missing stuff.  This all has to be tested.  Can these lines go in independently first in dunetpc, or do we have to build larsoft with the nine pull requests in order to even get something that runs with them?  I suspect a null config parameter is just fine to add by itself.

PetrilloAtWork commented:
Just noticed that `lariatsoft` is utterly outdated _and_ not in C.I. (for a reason). Stupid me.

PetrilloAtWork commented:
I think you need `larcoreobj`, `larcorealg` and `larcore`.
My number was the count of occurrences of `Geometry:`. Your method might be more reliable.

PetrilloAtWork commented:
I have verified that @tomjunk estimation is correct. With my keyword `Geometry:` I had counted configurations that imported a configuration and then overrode only `Geometry`, in which case no change is needed as long as the imported configuration is fixed elsewhere.

tomjunk commented:
I pushed a feature branch to dunetpc called **feature/trj_forgianluca_geo_dec2020**

It has GeometryConfigurationWriter: {} after each ExptGeometryHelperInterface instance in the fcl files, including the commented-out ones so I could count to see if I got them all.  I ran the protodune refactored prod4 sim+reco interactively with one beam+cosmics event and was able to view it in the event display.

PetrilloAtWork commented:
Following @tomjunk hint, I have modified most of the FHiCL configuration files containing `ExptGeoHelperInterface`, explicitly excluding the ones that appear to be obsolete.
Since my working area is not able to run all the tests (mostly because the various experiment repositories end up producing some duplicate ROOT dictionaries), I will need to figure out how to trigger a continuous integration test.
The branches offered for test, with the recommendation that the experiments check them, are all named `feature/gp_issue24328`, and are available for the following repositories:
* ArgoNeuT: `argoneutcode`
* DUNE: `dunetpc` (this is Tom Junk's very branch, rebranded)
* ICARUS: `sbncode`, `icarusalg`, `icaruscode`
* SBND: `sbncode`, `sbndcode`
* MicroBooNE: `uboonecode`, `ubana`, `ubcrt`, `ubcv`, `ubevt`, `ublite`, `ubraw`, `ubreco`, `ubsim`

Missing is `protoduneana`. @tomjunk, is it possible for you to take care _also_ of that one?

lgarren commented:
Thank you!  I presume the ICARUS and SBND branches have been added as PRs in the respective repositories.  Do you have the PR numbers for them?

Instructions for triggering the CI by hand are on the lar_ci wiki.  I'm going to trigger a couple now.

lgarren commented:
Note that I made myself a script to help keep everything straight!  I've triggered the CI tests for argoneutcode, dunetpc, and sbndcode.
`kx509

source /cvmfs/fermilab.opensciencegrid.org/products/common/etc/setups.sh
setup lar_ci

trigger --build-delay 0 --cert /tmp/x509up_u1147 --version develop  --workflow dunetpc_wf --revisions "LArSoft/larcoreobj#9 LArSoft/larcorealg#12 LArSoft/larcore#6 LArSoft/lardataalg#18 LArSoft/lardata#12 LArSoft/larevt#10 LArSoft/larsim#56 LArSoft/larreco#24 LArSoft/larexamples#5 dunetpc@feature/gp_issue24328"

trigger --build-delay 0 --cert /tmp/x509up_u1147 --version develop  --workflow argoneutcode_wf --revisions "LArSoft/larcoreobj#9 LArSoft/larcorealg#12 LArSoft/larcore#6 LArSoft/lardataalg#18 LArSoft/lardata#12 LArSoft/larevt#10 LArSoft/larsim#56 LArSoft/larreco#24 LArSoft/larexamples#5 argoneutcode@feature/gp_issue24328"

trigger --build-delay 0 --cert /tmp/x509up_u1147 --version develop  --workflow uboonecode_wf --revisions "LArSoft/larcoreobj#9 LArSoft/larcorealg#12 LArSoft/larcore#6 LArSoft/lardataalg#18 LArSoft/lardata#12 LArSoft/larevt#10 LArSoft/larsim#56 LArSoft/larreco#24 LArSoft/larexamples#5 uboonecode@feature/gp_issue24328 ubana@feature/gp_issue24328 ubcrt@feature/gp_issue24328 ubcv@feature/gp_issue24328 ubevt@feature/gp_issue24328 ublite@feature/gp_issue24328 ubraw@feature/gp_issue24328 ubreco@feature/gp_issue24328 ubsim@feature/gp_issue24328"
`

lgarren commented:
I forgot that SBNSoftware allows users to create feature branches.  ICARUS and SBND tests now triggered.

`trigger --build-delay 0 --cert /tmp/x509up_u1147 --version develop  --workflow sbndcode_wf --revisions "LArSoft/larcoreobj#9 LArSoft/larcorealg#12 LArSoft/larcore#6 LArSoft/lardataalg#18 LArSoft/lardata#12 LArSoft/larevt#10 LArSoft/larsim#56 LArSoft/larreco#24 LArSoft/larexamples#5 sbncode@feature/gp_issue24328 sbndcode@feature/gp_issue24328"

trigger --build-delay 0 --cert /tmp/x509up_u1147 --version develop  --workflow icaruscode_wf --revisions "LArSoft/larcoreobj#9 LArSoft/larcorealg#12 LArSoft/larcore#6 LArSoft/lardataalg#18 LArSoft/lardata#12 LArSoft/larevt#10 LArSoft/larsim#56 LArSoft/larreco#24 LArSoft/larexamples#5 icaruscode@feature/gp_issue24328 icarusalg@feature/gp_issue24328 sbncode@feature/gp_issue24328"
`

PetrilloAtWork commented:
Yep... sorry for the delayed answer.
I have produced no pull requests yet outside LArSoft, for the reason you mentioned.
If at any time a pull request is actually needed, just let me know.

lgarren commented:
@PetrilloAtWork  Alas, the argoneut, dune, and uboone CI builds are failing at the experiment build stage because your PR's are now behind the head of develop.  Would you please merge the head of develop with your PR's?  At minimum, the changes in ups/product_deps are required.

I note that the icarus and sbnd CI builds seem to have failed for undermined reasons.

lgarren commented:
@PetrilloAtWork  I've updated the argoneutcode, dunetpc, and uboone feature branches and restarted the CI.  Also, it appears that the correct syntax for sbn, sbnd, and icarus feature branches is SBNSoftware/sbncode@feature/gp_issue24328, so I'm restarting them with the correct syntax.

lgarren commented:
argoneutcode has completed with a warning about different product names
https://dbweb0.fnal.gov/LarCI/app/ns:ArgoNeuT/view_builds/index?offset=0&builds=argoneut_ci/5361&builds=

lgarren commented:
dunetpc results have warnings everywhere and one error
https://dbweb8.fnal.gov:8443/LarCI/app/ns:DUNE/view_builds/index?offset=0&builds=dune_ci/10642&builds=

lgarren commented:
uboone tests have warnings about product name changs
https://dbweb8.fnal.gov:8443/LarCI/app/ns:uBooNE/view_builds/index?offset=0&builds=uboone_ci/5781&builds=

PetrilloAtWork commented:
Lynn, thank you so much for pushing this forward: I have spent hours today violently quarreling with GIT to get to the same point you were five hours ago.
My comments on the reports:
* the `argoneutcode` warnings are correct and expected: there is a new data product
* the error in `dunetpc` code is caused by my change. I think the behaviour is correct, that the input file and the configuration of `Geometry` service of the test appear incompatible (different `DetectorName`). @tomjunk, are you able to comment on the failure? The relevant log appears to be [the error log](https://dbweb5.fnal.gov:8443/LarCI/app/ns:dune/storage/docs/2020/12/15/stderr.log) at https://dbweb8.fnal.gov:8443/LarCI/app/ns:DUNE/build_detail/test_details?build_id=dune_ci/10642&platform=Linux%203.10.0-1160.2.2.el7.x86_64&buildtype=slf7%20e19:prof&phase=ci_tests&test=ci_select_regression_test_dunefd ; the rest of the warnings look of the same flavour as for ArgoNeuT, and expected;
* MicroBooNE warnings appear to be a mix of [preexisting ones](https://dbweb8.fnal.gov:8443/LarCI/app/ns:uBooNE/build_detail/phase_details?build_id=uboone_ci/5779&platform=Linux%203.10.0-1127.19.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e19:prof) plus the ones equivalent to ArgoNeuT tests and expected.

From my point of view, these tests are good. I wonder if people at DUNE agree ;-)

PetrilloAtWork commented:
I have retriggered the integration tests for `sbndcode` and `icaruscode` with branches updated to LArSoft `v09_10_02`.
We'll see.

I am adding also a pull request LArSoft/lareventdisplay#10 . No test exercises anything in there, but for completeness it's there anyway.

lgarren commented:
@PetrilloAtWork Thanks for the sbnd and icarus updates. Unfortunately you used my first prescription, which is wrong.  My apologies.  I've triggered the build again.
`trigger --build-delay 0 --cert /tmp/x509up_u1147 --version develop  --workflow sbndcode_wf --revisions "LArSoft/larcoreobj#9 LArSoft/larcorealg#12 LArSoft/larcore#6 LArSoft/lardataalg#18 LArSoft/lardata#12 LArSoft/larevt#10 LArSoft/larsim#56 LArSoft/larreco#24 LArSoft/larexamples#5 SBNSoftware/sbncode@feature/gp_issue24328 SBNSoftware/sbndcode@feature/gp_issue24328"

trigger --build-delay 0 --cert /tmp/x509up_u1147 --version develop  --workflow icaruscode_wf --revisions "LArSoft/larcoreobj#9 LArSoft/larcorealg#12 LArSoft/larcore#6 LArSoft/lardataalg#18 LArSoft/lardata#12 LArSoft/larevt#10 LArSoft/larsim#56 LArSoft/larreco#24 LArSoft/larexamples#5 SBNSoftware/icaruscode@feature/gp_issue24328 SBNSoftware/icarusalg@feature/gp_issue24328 SBNSoftware/sbncode@feature/gp_issue24328"
`

lgarren commented:
@tomjunk are you happy with the dunetpc results?  (See the comment thread)

lgarren commented:
There is still a dependency discrepancy in icarusalg that is causing that the icarus CI tests to fail.   However, I can confirm that all unit tests pass for argoneut, dune, icarus, sbnd, and uboone.

The SBND CI tests report warnings.
https://dbweb8.fnal.gov:8443/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/5872&builds=

PetrilloAtWork commented:
Thank you again, Lynn. I noticed the issue with `icarusalg`, but apparently too late for the test.
The warnings from the `sbndcode` test appear to match the expected type.

I tend to recommend merging _after_ the holidays, but that may mean additional effort to catch up with the new changes in the FHiCL files (let's not expect physicists to take massive holidays...). But I yield to LArSoft's decision.

tomjunk commented:
I looked at the CI output.  It looks like the config consistency check is doing its job and DUNE has two names for the same or very similar geometries.  "workspace" vs "1x2x6".  I'm checking with Tingjun about what to do.

tomjunk commented:
I pushed a band-aid to ci_test_select_dunefd.fcl in dunetpc's develop just now to set the geometry GDML, Name and ROOT spec to match that of the input file in the CI test.  This looks like something that might be endemic in dunetpc -- the use of "compatible" geometries that now cause exceptions.

PetrilloAtWork commented:
So far, the only allowance to that situation is to configure the `Geometry` service to _ignore_ the result of the check (`SkipConfigurationCheck`, see the top post).
I consider the need to enable that option a symptom of an issue; you can assess if the issue is in DUNE configuration/customs or in the design/concept of the new check, and we can proceed from there.
